src/main/java/uk/ac/cam/cl/dtg/isaac/api/managers/EventBookingManager.java

```diff
@@ -899,12 +1160,17 @@ private void ensureValidBooking(final IsaacEventPageDTO event, final RegisteredU
         if (enforceBookingDeadline && event.getBookingDeadline() != null && now.after(event.getBookingDeadline())) {
             throw new EventDeadlineException("The booking deadline has passed.");
         }
-
+        // check if already reserved
+        if (requestedBookingStatus.equals(BookingStatus.RESERVED) && this.isUserReserved(event.getId(), user.getId())) {
+            throw new DuplicateBookingException(String.format("Unable to reserve onto event (%s) as user (%s) is"
+                    + " already reserved on to it.", event.getId(), user.getEmail()));
+        }
         // check if already booked
-        if (this.isUserBooked(event.getId(), user.getId())) {
+        if (requestedBookingStatus.equals(BookingStatus.CONFIRMED) && this.isUserBooked(event.getId(), user.getId())) {
             throw new DuplicateBookingException(String.format("Unable to book onto event (%s) as user (%s) is already"
                     + " booked on to it.", event.getId(), user.getEmail()));
         }
+        // should we check if already on waiting list too?
```

Comment:
Unresolved question

Reply:
Does a7f0a5ff answer this question? I think this is what I was thinking of when I asked it, but it's been a long time…

Reply:
Or maybe this is better, actually c6f25866.

Reply:
It doesn't answer the question as it doesn't justify why those changes were made.
I guess the question is:
if a user is on the waiting list, should it be possible to create a reservation for that user.
It should not be possible to create a new reserved booking because they both count towards the capacity of the event.
You probably also don't want to be updating from Waiting List to Reserved because reserved implies that the student needs to go and fill in their details where as if they are on the waiting list they will have already provided that.

Reply:
I'm going assume it's NOT OK to go from waiting list to reservation because that would allow a teacher (and/or group manager?) to make a student "jump the queue".

It will not be possible to create both a reservation and a waiting list booking precisely because of double counting.

I'm going to review the flow of that method with these assumptions.

Reply:
Reservation = take up one space but student needs to add more information
Waiting List = take up one space and student has provided all necessary information

So we wouldn't want to change a booking from waiting list to a reservation.
This would actually only be a possible if there are say 8 bookings for a 5 space event, leaving 3 on the waiting list, then the 5 confirmed cancelled. The teacher could then try to reserve one of the 3 on the waiting list into the 2 available spaces. As it is possible we should cover the case.

Reply:
OK, now. db5daf08 should clarify this a bit more. Hopefully that's the last of it.